File: 04_BehaviouralDesignPatterns/01_IteratorPattern/cpp/Main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <memory>

class Video {
private:
    std::string title;

public:
    Video(const std::string& title) : title(title) {}

    std::string getTitle() const {
        return title;
    }
};

// Iterator interface
class Iterator {
public:
    virtual bool hasNext() = 0;
    virtual Video* next() = 0;
    virtual ~Iterator() = default;
};

// Collection interface
class Collection {
public:
    virtual Iterator* createIterator() = 0;
    virtual ~Collection() = default;
};

// Concrete Iterator
class PlaylistIterator : public Iterator {
private:
    std::vector<Video>& videos;
    size_t position;

public:
    PlaylistIterator(std::vector<Video>& videos) : videos(videos), position(0) {}

    bool hasNext() override {
        return position < videos.size();
    }

    Video* next() override {
        if (hasNext()) {
            return &videos[position++];
        }
        return nullptr;
    }
};

// Concrete Collection
class YouTubePlaylist : public Collection {
private:
    std::vector<Video> videos;

public:
    void addVideo(const Video& video) {
        videos.push_back(video);
    }

    Iterator* createIterator() override {
        return new PlaylistIterator(videos);
    }
};

int main() {
    YouTubePlaylist playlist;
    playlist.addVideo(Video("Video 1"));
    playlist.addVideo(Video("Video 2"));
    playlist.addVideo(Video("Video 3"));

    // Using iterator to access videos
    std::unique_ptr<Iterator> iterator(playlist.createIterator());
    while (iterator->hasNext()) {
        Video* video = iterator->next();
        std::cout << video->getTitle() << std::endl;
    }

    return 0;
}

File: 04_BehaviouralDesignPatterns/01_IteratorPattern/Main.java
import java.util.ArrayList;
import java.util.List;

class Video {
    String title;

    public Video(String title) {
        this.title = title;
    }

    public String getTitle() {
        return title;
    }
}

// class YoutTubePlaylist {
//     private List<Video> videos = new ArrayList<>();

//     public void addVideo(Video video) {
//         videos.add(video);
//     }

//     public List<Video> getVideos() {
//         return videos;
//     }
// }

interface PlaylistIterator {
    boolean hasNext();
    Video next();
}

class YoutTubePlaylistIterator implements PlaylistIterator {
    private List<Video> videos;
    private int position;

    public YoutTubePlaylistIterator(List<Video> videos) {
        this.videos = videos;
        this.position = 0;
    }

    @Override
    public boolean hasNext() {
        return position < videos.size();
    }

    @Override
    public Video next() {
        if (!hasNext()) {
            throw new IndexOutOfBoundsException("No more videos in the playlist");
        }
        return videos.get(position++);
    }
}

// Iterable interface
interface Playlist {
    PlaylistIterator createIterator();
    PlaylistIterator createCopyrightIterator();
}

class YoutTubePlaylist implements Playlist {
    private List<Video> videos = new ArrayList<>();

    public void addVideo(Video video) {
        videos.add(video);
    }

    @Override
    public PlaylistIterator createIterator() {
        return new YoutTubePlaylistIterator(videos);
    }

    @Override
    public PlaylistIterator createCopyrightIterator() {
        // This method can be implemented to return a different iterator if needed
        return new YoutTubePlaylistIterator(videos);
    }
}

public class Main {
    public static void main(String[] args) {
        YoutTubePlaylist playlist = new YoutTubePlaylist();
        playlist.addVideo(new Video("Video 1"));
        playlist.addVideo(new Video("Video 2"));
        playlist.addVideo(new Video("Video 3"));
        playlist.addVideo(new Video("Video 4"));
        playlist.addVideo(new Video("Video 5"));

        PlaylistIterator iterator = playlist.createIterator();

        while (iterator.hasNext()) {
            Video video = iterator.next();
            System.out.println(video.getTitle());
        }
    }
}